File: backend/src/services/mqttService.js
const mqtt = require('mqtt');
const db = require('../config/database');
const logger = require('../config/logger');

const MQTT_BROKER = process.env.MQTT_BROKER || 'mqtt://localhost:1883';
const TOPIC = 'campuseats/scales/#';

let client;

const start = () => {
  client = mqtt.connect(MQTT_BROKER, {
    clientId: `campuseats-server-${Date.now()}`,
    reconnectPeriod: 5000,
  });

  client.on('connect', () => {
    logger.info(`MQTT connected to ${MQTT_BROKER}`);
    client.subscribe(TOPIC, (err) => {
      if (err) logger.error('MQTT subscribe error', err);
    });
  });

  client.on('message', async (topic, payload) => {
    try {
      // Topic format: campuseats/scales/<device_id>
      const device_id = topic.split('/')[2];
      const data = JSON.parse(payload.toString());
      const { weight_kg, cafeteria_id } = data;

      if (typeof weight_kg !== 'number' || weight_kg < 0) {
        logger.warn(`Invalid payload from ${device_id}`, data);
        return;
      }

      await db.query(
        `INSERT INTO iot_readings (device_id, cafeteria_id, weight_kg, raw_payload)
         VALUES ($1, $2, $3, $4)`,
        [device_id, cafeteria_id, weight_kg, data]
      );

      // Anomaly detection: alert if single reading > 5 kg
      if (weight_kg > 5) {
        await db.query(
          `INSERT INTO alerts (cafeteria_id, type, severity, message)
           VALUES ($1, 'spike', 'high', $2)`,
          [cafeteria_id, `Scale ${device_id} reported ${weight_kg} kg — possible waste spike`]
        );
        logger.warn(`Waste spike alert: ${device_id} ${weight_kg} kg`);
      }
    } catch (err) {
      logger.error('MQTT message processing error', err);
    }
  });

  client.on('error', (err) => logger.error('MQTT error', err));
  client.on('reconnect', () => logger.info('MQTT reconnecting…'));
};

const stop = () => client?.end();

module.exports = { start, stop };

File: backend/services/mqtt_service.py
import asyncio
import json
import logging

import aiomqtt

from core.config import settings
from core.database import get_pool

logger = logging.getLogger(__name__)
TOPIC = "campuseats/scales/#"


async def handle_message(message: aiomqtt.Message) -> None:
    try:
        device_id = str(message.topic).split("/")[2]
        data = json.loads(message.payload)
        weight_kg = data.get("weight_kg")
        cafeteria_id = data.get("cafeteria_id")

        if not isinstance(weight_kg, (int, float)) or weight_kg < 0:
            logger.warning("Invalid payload from %s: %s", device_id, data)
            return

        pool = await get_pool()
        async with pool.acquire() as conn:
            await conn.execute(
                "INSERT INTO iot_readings (device_id, cafeteria_id, weight_kg, raw_payload) VALUES ($1,$2,$3,$4)",
                device_id, cafeteria_id, weight_kg, json.dumps(data),
            )
            if weight_kg > 5:
                await conn.execute(
                    "INSERT INTO alerts (cafeteria_id, type, severity, message) VALUES ($1,'spike','high',$2)",
                    cafeteria_id,
                    f"Scale {device_id} reported {weight_kg} kg — possible waste spike",
                )
                logger.warning("Waste spike alert: %s %.2f kg", device_id, weight_kg)
    except Exception:
        logger.exception("Error processing MQTT message")


async def run_mqtt_listener() -> None:
    while True:
        try:
            async with aiomqtt.Client(
                hostname=settings.MQTT_BROKER_HOST,
                port=settings.MQTT_BROKER_PORT,
            ) as client:
                logger.info("MQTT connected to %s:%s", settings.MQTT_BROKER_HOST, settings.MQTT_BROKER_PORT)
                await client.subscribe(TOPIC)
                async for message in client.messages:
                    await handle_message(message)
        except Exception:
            logger.exception("MQTT error, reconnecting in 5s…")
            await asyncio.sleep(5)
